auth-service/src/main/kotlin/uk/gov/communities/delta/auth/services/UserService.kt

```diff
@@ -10,31 +15,101 @@ import javax.naming.directory.*
 class UserService(
     private val ldapServiceUserBind: LdapServiceUserBind,
     private val userLookupService: UserLookupService,
+    private val userAuditService: UserAuditService,
 ) {
 
     private val logger = LoggerFactory.getLogger(javaClass)
 
-    suspend fun createUser(adUser: ADUser) {
+    suspend fun createUser(
+        adUser: ADUser,
+        ssoClient: AzureADSSOClient?,
+        triggeringAdminSession: OAuthSession?,
+        call: ApplicationCall,
+        azureUserObjectId: String? = null,
+    ) {
         try {
             addUserToAD(adUser)
         } catch (e: Exception) {
             logger.atError().addKeyValue("UserDN", adUser.dn).log("Error creating user", e)
             throw e
         }
+        auditUserCreation(adUser.cn, ssoClient, triggeringAdminSession, call, getAuditData(adUser), azureUserObjectId)
+    }
+
+    private suspend fun auditUserCreation(
+        userCN: String,
+        ssoClient: AzureADSSOClient?,
+        triggeringAdminSession: OAuthSession?,
+        call: ApplicationCall,
+        auditData: MutableMap<String, String>,
+        azureUserObjectId: String?,
+    ) {
+        val ssoUser = ssoClient?.required == true
+        if (ssoUser) {
+            auditData["ssoClientInternalId"] = ssoClient!!.internalId
+            if (azureUserObjectId != null) auditData["azureObjectId"] = azureUserObjectId
+        }
+        val encodedAuditData = Json.encodeToString(auditData)
+        if (triggeringAdminSession != null) {
+            if (ssoUser) userAuditService.ssoUserCreatedByAdminAudit(
+                userCN,
+                triggeringAdminSession.userCn,
+                call,
+                encodedAuditData
+            )
+            else userAuditService.userCreatedByAdminAudit(userCN, triggeringAdminSession.userCn, call, encodedAuditData)
+        } else if (ssoUser) userAuditService.userCreatedBySSOAudit(
+            userCN,
+            call,
+            encodedAuditData
+        ) else userAuditService.userSelfRegisterAudit(userCN, call, encodedAuditData)
+    }
+
+    private fun getAuditData(adUser: ADUser): MutableMap<String, String> {
+        val auditData = mutableMapOf<String, String>()
+        auditData["userPrincipalName"] = adUser.userPrincipalName
+        auditData["cn"] = adUser.cn
+        auditData["sn"] = adUser.sn
+        auditData["givenName"] = adUser.givenName
+        auditData["mail"] = adUser.mail
+        auditData["st"] = adUser.notificationStatus
+        auditData["userAccountControl"] = adUser.userAccountControl
+
+        if (adUser.comment != null) auditData["comment"] = adUser.comment!!
```

Comment:
It's annoying we have to use `!!`. Could do `adUser.comment?.let { auditData["comment"] = it }` but up to you whether you think that's clearer